### ISSUE TITLE
Addition of dual Serial/MPI builds

### ISSUE THREAD

**Issue by lexming** [easybuilders/easybuild#649]
Following the conf call on [dual serial/MPI builds](https://github.com/easybuilders/easybuild/wiki/Conference-Call-on-Dual-Serial-MPI-builds), we now have a clearer path forward with this project.

Any actions taken will be tracked in the projects page: https://github.com/orgs/easybuilders/projects/8

First wave of libraries to be dualized:
- [ ] HDF5 (https://github.com/easybuilders/easybuild-easyconfigs/pull/11145)
- [ ] Boost (https://github.com/easybuilders/easybuild-easyconfigs/pull/11155)
- [ ] FFTW
- [ ] VTK

Anybody interested in contributing to this project can do so by:
- Making PRs and testing the serial and MPI builds of the aforementioned packages
- Searching, reporting or PRing easyconfigs that could benefit from switching to serial dependencies. Please follow these requirements
  - Only easyconfigs in the EasyBuild tree
  - Only easyconfigs in the latests toolchains (`foss/2020a` or `intel/2020a` and compatible toolchains)

**Comment by Micket** [comment]
So, an important point to resolve between this can proceed is how compatible an HDF5.Serial and HDF5(.MPI) is. Because one shouldn't be allow to break everything by simply loading 2 modules normally, which we can boil down to, e.g.
```
module load TensorFlow    # will load HDF5 (MPI) as a dependency (via h5py)
module load SRA-Toolkit   # will load HDF5.Serial as a dependency

mpirun python my_machine_learning_project.py
```
can't be allowed to result in anything broken, and we can't assume use of RPATH. If this leads to anything broken (and I strongly suspect it does, because.. how could it not?) then we must use a versionsuffix (e.g. `HDF5-1.12.0-gompi-2020a-serial.eb`) instead to ensure a package conflict,  instead of pretending that it is compatible and quietly break things.

**Comment by lexming** [comment]
@Micket I totally agree, that is why we will work in the `dual_serial_mpi` branch for now. To test such cases and see where things break. As soon as we hit any insurmountable issues, then we can revise our strategy.

**Comment by smoors** [comment]
the valid concerns raised by @Micket may be (partially?) mitigated if the following rules are adopted:
* prerequisite: MPI version is superset of serial version, that is, all symbols in serial version are also in MPI version
* rule 1) MPI version depends on serial version
* rule 2) serial version does not depend on MPI version
* rule 3) serial version is not loaded when MPI version is loaded, preventing that serial version jumps before MPI version in LD_LIBRARY_PATH

possible implementation of rule 3) in Lmod:
```
if not ( isloaded("name.MPI") ) then
    <set environment for name.serial>
end
```

possible implementation of rule 3) in Tcl:
```
if { ![ is-loaded name.MPI ] } {
    <set environment for name.serial>
}
```

expected results:
* case 1: loading only MPI version   
due to 1), serial version is loaded first and MPI version last, thus libs of MPI version come first in LD_LIBRARY_PATH
* case 2: loading only serial version
due to 2), only serial version is in LD_LIBRARY_PATH
* case 3: loading MPI version first, serial version second
due to 3), only MPI version is in LD_LIBRARY_PATH
* case 4: loading serial version first, MPI version second
due to order of loading, libs of MPI version come first in LD_LIBRARY_PATH
